poseidon/build.rs

```diff
@@ -0,0 +1,349 @@
+            Sign::Minus => -self.digits.parse::<i64>().unwrap(),
+        }
+    }
+
+    fn usize(&self) -> usize {
+        assert_eq!(self.sign, Sign::Plus);
+        self.digits.parse().unwrap()
+    }
+
+    fn field<F: PrimeField>(&self) -> F {
+        assert_eq!(self.sign, Sign::Plus);
+        let mut res = F::ZERO;
+        for c in self.digits.chars() {
+            let digit = match c {
+                '0' => F::from(0u64),
+                '1' => F::from(1u64),
+                '2' => F::from(2u64),
+                '3' => F::from(3u64),
+                '4' => F::from(4u64),
+                '5' => F::from(5u64),
+                '6' => F::from(6u64),
+                '7' => F::from(7u64),
+                '8' => F::from(8u64),
+                '9' => F::from(9u64),
+                _ => panic!("Invalid digit"),
+            };
+            res = res * F::from(10u64) + digit;
+        }
+        res
+    }
+}
+
+fn parse() -> impl Parser<char, Vec<Options>, Error = Simple<char>> {
+    fn opt(c: char) -> impl Parser<char, bool, Error = Simple<char>> + Clone {
+        choice((just(c).map(|_| true), empty().map(|_| false)))
+    }
+
+    let num = opt('-')
+        .then(text::int(10))
+        .map(|(neg, digits)| Number {
+            sign: if neg { Sign::Minus } else { Sign::Plus },
+            digits,
+        })
+        .padded();
+
+    let row = num.clone().separated_by(just(',').padded()).delimited_by(
+        just('[').padded(),
+        opt(',').padded().then(just(']')).padded(),
+    );
+
+    let mxt = row
+        .separated_by(just(',').padded())
+        .delimited_by(
+            just('[').padded(),
+            opt(',').padded().then(just(']')).padded(),
+        )
+        .padded();
+
+    let eq = just("=").padded();
+
+    let opt = choice((
+        // parse the rate
+        just("Rate")
+            .then(eq)
+            .then(num.clone())
+            .map(|(_, num)| Options::Rate(num)),
+        // parse the capacity
+        just("Capacity")
+            .then(eq)
+            .then(num.clone())
+            .map(|(_, num)| Options::Capacity(num)),
+        // parse the number of full rounds
+        just("FullRounds")
+            .then(eq)
+            .then(num.clone())
+            .map(|(_, num)| Options::FullRounds(num)),
+        // parse the number of partial rounds
+        just("PartialRounds")
+            .then(eq)
+            .then(num.clone())
+            .map(|(_, num)| Options::PartialRounds(num)),
+        // parse the matrix
+        just("MDS")
+            .then(eq)
+            .then(mxt.clone())
+            .map(|(_, m)| Options::Mds(m)),
+        // parse the round keys list
+        just("RoundKeys")
+            .then(eq)
+            .then(mxt.clone())
+            .map(|(_, m)| Options::RoundKeys(m)),
+    ));
+
+    opt.padded().repeated().then(end()).map(|(v, _)| v)
+}
+
+#[allow(clippy::single_char_add_str)]
+fn main() {
```

Comment:
Is there any benefit to including builder code rather than pre-built code?
Or it's just a matter of choice? I'm just curios 😅

Reply:
I went back/forth on this: we could parse them once and include the compressed round keys.

- On one hand it would simplify things: just have an array with the processed round keys.
- On the other hand with would make it less clear where these constants come from, if, e.g. I had a separate python script producing them and then simply added these as constants.

It might be over engineered: roughly half the code is doing this.

Reply:
Your right, I decided that it was simpler to just add the processed output.